File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <stack>
#include <string>

using namespace std;

bool check(string str) {
	int length = str.length();
	stack<char> st;
	
	for (int i=0; i<length; i++) {
		char ch = str[i];
		
		if(ch == '(') {
			st.push(ch);
		}
		else {
			if(!st.empty()) {
				st.pop();
			}
			else {
				return false;
			}
		}
	}
	
	return st.empty();
}

int main() {
	
	int n;
	cin >> n;
	
	for(int i =0; i < n; i++) {
		string str;
		cin >> str;
		
		if(check(str)) {
			cout << "YES" << '\n';
		}
		else {
			cout << "NO" << '\n';
		}
	}
	
	return 0;
}

File: 백준/Silver/9012. 괄호/괄호.java
import java.util.Scanner;
import java.util.Stack;

public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int next = sc.nextInt();
        sc.nextLine();

        while(next-- > 0) {
            String line = sc.nextLine();
            System.out.println(solution(line));
        }
    }


    public static String solution(String input) {

        Stack<Integer> pStack = new Stack<>();

        for (char p : input.toCharArray()) {
            if(p == ')') {
                if(pStack.isEmpty())
                    return "NO";
                else
                    pStack.pop();
            }

            else if(p == '(') {
                pStack.push(1);
            }
        }

        if(pStack.isEmpty()) return "YES";
        return "NO";
    }
}